### PR TITLE
[6.x] Increase performance of RefreshDatabase trait

### PR DESCRIPTION
**Pull request by sileence**
This PR allows `RefreshDatabase` to cache the last modified migration time, in order to avoid re-running the migrations locally if no migration files have been added or changed. 

The first time I run a feature test it takes around 3.66 seconds mostly due to `migrate:fresh` being executed (it can be worse in projects with a lot of migration files). With the new version of this trait, the second time onwards I run the tests only take around 0.6 seconds. So 5x times faster.

This code has been tested in Linux and Mac with similar speed gain.

### CODE REVIEW

**Comment by netpok** [comment]
I'm using a similar approach in our systems, but this implementation does not work in most cases. If I saw it correctly this uses the cache and most of the time tests are using the array cache driver.

File: src/Illuminate/Foundation/Testing/DatabaseMigrations.php [hunk 1]
```diff
@@ -20,7 +20,7 @@ public function runDatabaseMigrations()
         $this->beforeApplicationDestroyed(function () {
             $this->artisan('migrate:rollback');
 
-            RefreshDatabaseState::$migrated = false;
+            (new DatabaseState($this->app))->markAsMigrated(false);
```
Comment (netpok) [inline src/Illuminate/Foundation/Testing/DatabaseMigrations.php]: You could use `$this->app[DatabaseState::class]->markAsMigrated(false);` here

File: src/Illuminate/Foundation/Testing/DatabaseState.php [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+     * @var string
+     */
+    const LAST_MODIFIED_KEY = 'database_last_modified_migration';
+
+    /**
+     * Flag to indicate if the test database has been migrated recently.
+     *
+     * @var bool
+     */
+    private static $recentlyMigrated = false;
```
Comment (netpok) [inline src/Illuminate/Foundation/Testing/DatabaseState.php]: Why private? And also there is no way to retrieve the value of this.

File: src/Illuminate/Foundation/Testing/RefreshDatabaseState.php [hunk 3]
```diff
@@ -1,13 +0,0 @@
-<?php
```
Comment (netpok) [inline src/Illuminate/Foundation/Testing/RefreshDatabaseState.php]: You shouldn't remove files in current version, some implementation can depend on it, it should target master.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Comment by sileence** [comment]
I've uploaded a new package with a similar and improved implementation of this idea for anyone who might be interested: https://github.com/StydeNet/laravel-test-helpers

**Pull request closed without merging**